### PR TITLE
Enrich keycloak SessionState with name+username

### PR DESCRIPTION
**Pull request by bodom0015**
Enrich the session state with additional user info to consume within client applications.

This way, the user can call the `/userinfo` endpoint with a valid token to retrieve these attributes and consume them within client applications to create any server-side resources necessary to create a new user account.

## Description

There was already code to [enrich the session with the user's email from keycloak](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/keycloak.go#L91-L95), so it was not much additional code to also add the user's `name` and `preferred_username`

## Motivation and Context

This will address part of #834 (for the Keycloak case). Further changes may be needed for other providers

## How Has This Been Tested?

Manually tested with Keycloak + CILogon integration.

Since the change should only affect the keycloak provider, I did only test under configurations where that provider would be used. I have not thoroughly tested or exercised other providers.

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

Is it ok to push a new feature branch to this repo? I am not already a collaborator, so I'm not sure if I have sufficient access

### CODE REVIEW

File: providers/keycloak.go [hunk 1]
```diff
@@ -94,5 +94,17 @@ func (p *KeycloakProvider) EnrichSession(ctx context.Context, s *sessions.Sessio
 	}
 	s.Email = email
 
+	fullName, err := json.Get("name").String()
+	if err != nil {
+		return fmt.Errorf("unable to extract name from userinfo endpoint: %v", err)
+	}
+	s.User = fullName
+
+	username, err := json.Get("preferred_username").String()
```
Comment (NickMeves) [inline providers/keycloak.go:103]: We are moving to deprecate this -- lets not add it.

File: providers/keycloak.go [hunk 2]
```diff
@@ -94,5 +94,17 @@ func (p *KeycloakProvider) EnrichSession(ctx context.Context, s *sessions.Sessio
 	}
 	s.Email = email
 
+	fullName, err := json.Get("name").String()
```
Comment (NickMeves) [inline providers/keycloak.go:97]: What does Keycloak get now as the `session.User`? In other OIDC based providers, it is extracted from the ID Token subject claim (but this Keycloak provider is incorrectly using OAuth even though it should use OIDC).

**Comment by NickMeves** [comment]
So we actually want to kill this Provider soon and replace it with a `keycloak-oidc` (that will eventually become `keycloak` after a deprecation warning and major release): https://github.com/oauth2-proxy/oauth2-proxy/issues/956

If you are looking to use Keycloak & have bandwidth to make some improvements, I'd love help making the rearchitecture in that issue.

I think initially before adding more functionality (Keycloak roles) -- it might just be a simple Provider where in its struct it embeds the `OIDCProvider` and just uses all its implementation. You can see an ADFS PR using that technique here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1021/files

File: providers/keycloak.go (hunk 2 above)
Comment (bodom0015) [reply providers/keycloak.go:97]: Unclear what the internal values were for `session.User`, as I wasn't getting this a value for this field back from the `/userinfo` endpoint at all. 

Could this have been affected by the mappings in Keycloak for my particular OIDC provider (CILogon)? I'm a bit fuzzy on how all of these scopes and mappings actually connect with the `/userinfo` produced, so I'm not sure how to answer your question.

Comment (NickMeves) [reply providers/keycloak.go:97]: FYI - here's a draft PR we are working through to completely replace the existing Keycloak provider: https://github.com/oauth2-proxy/oauth2-proxy/pull/1107

**Pull request closed without merging**